### PR TITLE
Sprint 1

### PR DESCRIPTION
**Pull request by PaddyEriksen**

### CODE REVIEW

File: Analytics/app.py [hunk 1]
```diff
@@ -1,38 +1,144 @@
...
+from resources.Widgets.save_layouts import SaveWidgetLayout
+
+from resources.admin.create_new_user import CreateNewUser
+from resources.admin.user_permissions import UserPermissions
+from resources.admin.user_list import UsersList
+from resources.admin.change_user_password import ChangeUserPassword
+from resources.admin.delete_user import DeleteUser
+from resources.admin.change_user_name import ChangeUserName
+from resources.admin.get_user import GetUserByEmail
+from resources.admin.edit_user import EditUser
```
Comment (j0rdanl0urens) [inline Analytics/app.py]: Please sort out the imports

Comment (j0rdanl0urens) [reply Analytics/app.py]: python imports

3rd party imports

local imports

File: .gitignore [hunk 2]
```diff
@@ -1,2 +1,4 @@
 .DS_Store
-.idea/
\ No newline at end of file
+.idea/
+*.pyc
+*.pyc
```
Comment (DannyLee12) [inline .gitignore]: Only once please

File: Analytics/app.py [hunk 3]
```diff
@@ -1,38 +1,140 @@
...
+        """ Add admin claim to access token
+            :param user: Users model
+            :type user: Users instance
+            :return: Admin claim to be added to access JWT
+            :rtype: JSON
+        """
+        return {'admin': user.admin}
+
+    @jwt.user_identity_loader
+    def user_identity_lookup(user):
```
Comment (kpennels) [inline Analytics/app.py]: `def user_identity_lookup(user: Users) -> str:`

File: Analytics/models/widget.py [hunk 4]
```diff
@@ -44,6 +45,7 @@ def __init__(self, user_id, layout, data):
 
         # Does the database table exist?
         self.create_table()
+        logging.basicConfig(filename="widget_event.log", level=logging.DEBUG)
```
Comment (DannyLee12) [inline Analytics/models/widget.py]: This should be after the imports.

```python
logging.basicConfig(level='INFO')
logger = logging.getLogger(__name__)
```

Comment (DannyLee12) [reply Analytics/models/widget.py]: Then use 
```python
logger.debug("")
logger.info("info message")
logger.warning("")
logger.error("")
logger.critical("")
```

File: Analytics/models/widget.py [hunk 5]
```diff
@@ -81,7 +81,7 @@ def save(self) -> NoReturn:
             db.session.flush()
         except IntegrityError as ie:
             db.session.rollback()
-            logging.info(ie)
+            logging.error(ie)
```
Comment (DannyLee12) [inline Analytics/models/widget.py]: This should be 
```python
logger.error(ie)
```

File: Analytics/resources/admin/test_user_creation.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+class UserCreationTests(unittest.TestCase):
+
+    def setUp(self):
+
+        self.fullname = "abcdefg hijklmnop"
+        self.email = "james.bond@email.com"
+        self.password = "password"
+
+        self.url = "http://0.0.0.0:8005/admin"
+        self.jwt_access_token = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpYXQiOjE1NTA0MDczMDUsIm5iZiI6MTU1MD" \
```
Comment (j0rdanl0urens) [inline Analytics/resources/admin/test_user_creation.py]: No plain text keys. Either make this fake or store it securly

File: Analytics/resources/change_user_password.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+from flask_restful import Resource
+from flask_restful import abort
+from flask_restful import reqparse
+from sqlalchemy import exc
+
+from models.users import Users
+
+
+class ChangeUserPassword(Resource):
+
```
Comment (j0rdanl0urens) [inline Analytics/resources/change_user_password.py]: Docstrings to file

File: Analytics/resources/changePassword.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        :return: A message that indicates whether a user's password has been updated. If they have not, the message indicates why not.
+        :rtype: JSON
+	"""
+	parser = reqparse.RequestParser()
+	parser.add_argument('email', type=str, store_missing=False, help = 'This field cannot be blank', required = True)
+	parser.add_argument('password_old', type=str, store_missing=False, help = 'This field cannot be blank', required = True)
+	parser.add_argument('password_new', type=str, store_missing=False, help= 'This field cannot be blank', required = True)
+
+	@jwt_required
+	def patch(self):
```
Comment (j0rdanl0urens) [inline Analytics/resources/changePassword.py]: Docs

File: Analytics/resources/login.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	parser.add_argument('password', type=str, store_missing=False, help= 'This field cannot be blank', required = True)
+
+	def post(self):
+		args = self.parser.parse_args()
+		current_user = Users.find_by_email(args['email']) #t
+
+		if current_user and Users.verify_hash(args['password'].encode("utf8"), current_user.password.encode("utf8")):
+
+			if current_user.activated:
+					access_token = create_access_token(identity = current_user)
```
Comment (j0rdanl0urens) [inline Analytics/resources/login.py]: These should not be double indented

File: Analytics/resources/logout.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+class UserLogoutAccess(Resource):
+    """
+    API that revokes a users access JWT by adding it to the revoked_tokens table
+    :required: A valid access JWT in the Authorization Header in the format - Bearer <JWT>
+    :return: A message that indicates whether the token has been revoked 
+    :rtype: JSON
+    """
+    
+    @jwt_required
+    def post(self):
```
Comment (j0rdanl0urens) [inline Analytics/resources/logout.py]: Docs

File: Analytics/resources/logout.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        jti = get_raw_jwt()['jti']
+        try:
+            revoked_token = RevokedTokens(jti = jti)
+            revoked_token.add()
+            return {'message': 'Access token has been revoked'}
```
Comment (j0rdanl0urens) [inline Analytics/resources/logout.py]: Shouldn't a status code be returned?

File: Analytics/resources/logout.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    :rtype: JSON
+    """
+    
+    @jwt_refresh_token_required
+    def post(self):
+        jti = get_raw_jwt()['jti']
+        try:
+            revoked_token = RevokedTokens(jti = jti)
+            revoked_token.add()
+            return {'message': 'Refresh token has been revoked'}
```
Comment (j0rdanl0urens) [inline Analytics/resources/logout.py]: See above

File: Analytics/resources/register.py [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        :rtype: JSON
+	"""     
+
+	parser = reqparse.RequestParser()
+	parser.add_argument('fullname', type=str, store_missing=False) 
+	parser.add_argument('email', type=str, store_missing=False, help = 'This field cannot be blank', required = True)
+	parser.add_argument('password', type=str, store_missing=False,  help = 'This field cannot be blank', required = True)
+	parser.add_argument('password_new', type=str, store_missing=False,  help = 'This field cannot be blank', required = True)
+
+	def post(self):
```
Comment (j0rdanl0urens) [inline Analytics/resources/register.py]: Docs

File: Analytics/resources/register.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        :param password: users password that was sent when they were added on the admin page
+        :type fullname: string
+        :type email: string
+        :type password: string
+        :return: A message that indicates whether a user has been registered. If they have not, the message indicates why not
+        :rtype: JSON
+	"""     
+
+	parser = reqparse.RequestParser()
+	parser.add_argument('fullname', type=str, store_missing=False)
```
Comment (j0rdanl0urens) [inline Analytics/resources/register.py]: Question: should this be required or is it fine?

File: Analytics/resources/test_login_logout.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+def dummy_user():
+	""" A fixture that creates and saves a user to the database that is to be used for testing purposes """
+	
+    user = Users("User","user@FCC.com",Users.generate_hash("wfnbqk".encode("utf8")).decode("utf8"),True,True)
+
+    user.save()
+    user.commit()
+
+    return user
+
```
Comment (j0rdanl0urens) [inline Analytics/resources/test_login_logout.py]: Remove new lines

File: Analytics/resources/test_login_logout.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def test_login_correct_details(test_client, dummy_user):
+	""" 
+	Tests whether the correct response message, response code and corresponding access and refresh tokens are
+	returned when the correct login credentials are supplied 
+	"""
```
Comment (j0rdanl0urens) [inline Analytics/resources/test_login_logout.py]: Params

File: Analytics/resources/admin/test_user_creation.py (hunk 6 above)
Comment (j0rdanl0urens) [reply Analytics/resources/admin/test_user_creation.py]: I was told this was a real access token

File: Analytics/models/users.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+    def json(self) -> dict:
+        """
+        Create a JSON dict of the User objects attributes
+        :return: the User objects attributes as a JSON (dict)
+        """
+<<<<<<< HEAD
+=======
+        
+>>>>>>> c7eb421928183f26805c9b66a4eea5ebeca6de48
```
[the remark this answers is not in the document]
Comment (jamesseth) [reply Analytics/models/users.py]: lines 85 to 88 inclusive removed from Analytics/models/users.py and committed

File: Analytics/app.py [hunk 18]
```diff
@@ -1,38 +1,138 @@
...
-    from models.location import Location
-    from models.unit import Unit
-    from models.sensor_attribute import SensorAttribute 
-    from models.theme import Theme, SubTheme
+
+    jwt = JWTManager(app)
+
+    @jwt.token_in_blacklist_loader
+    def check_if_token_in_blacklist(decrypted_token):
+        """ Query revoked tokens table for presence of decrypted_token argument
```
[the remark this answers is not in the document]
Comment (jamesseth) [reply Analytics/app.py]: Docstring indentation corrected from line 71 to 100

File: Analytics/models/widget_layout.py [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+        # Does the table exist?
+        has_table = db.engine.dialect.has_table(db.engine, 'layouts')
+        return has_table
+
+    @classmethod
+    def get_layout_by_widget_id(cls, widgetID: int) -> object:
+        """
+        fetch layout instance using the widgets id from database
+        :param widgetID: the identification number of the widget related to the layout to fetch
+        :type Integer:
```
[the remark this answers is not in the document]
Comment (jamesseth) [reply Analytics/models/widget_layout.py]: Editing doctring to only have the basic description of the functionallity of the function, let me know if its should be removed entirely.

File: Analytics/resources/Widgets/test_create_widget.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
+import pytest
+from models.users import Users
+from models.revoked_tokens import RevokedTokens
+from app import create_app
+from db import db
+from flask_jwt_extended import create_access_token
```
[the remark this answers is not in the document]
Comment (jamesseth) [reply Analytics/resources/Widgets/test_create_widget.py]: sorted out imports

File: Analytics/resources/test_login_logout.py (hunk 15 above)
Comment (jamesseth) [reply Analytics/resources/test_login_logout.py]: Fixed indentation and blank lines

File: Analytics/models/users.py [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+
+    id = db.Column(db.Integer, primary_key=True)
+    fullname = db.Column(db.String(400), nullable=False)
+    email = db.Column(db.String(120), unique=True, nullable=False)
+    password = db.Column(db.String(400), nullable=False)
+    admin = db.Column(db.Boolean)
+    activated = db.Column(db.Boolean)
+    timestamp = db.Column(db.DateTime)
+
+    def __init__(self, fullname, email, password, admin, activated, timestamp=None):
```
Comment (DannyLee12) [inline Analytics/models/users.py]: Please use python3 typehiints, 
```suggestion
    def __init__(self, fullname: str, email: str, password: str, admin: bool, activated: bool, timestamp: datetime.datetime=None):
```

File: Analytics/models/users.py [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+        """
+        Initialises the Users object instance
+        :param fullname:    users full name
+        :param email:       users email address
+        :param password:    plain text password
+        :param admin:       true if the user is a admin, false if the user is not an admin
+        :param activated:   true if the account has been activated and false if it is not activated
+        :param timestamp:   time stamp of when the user was created
+
+        :type fullname:     String
```
Comment (DannyLee12) [inline Analytics/models/users.py]: These type hints are invalid. 
String -> str
Boolean -> bool
DateTime -> datetime.datetime (depending on the import)

File: Analytics/models/revoked_tokens.py [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+    jti = db.Column(db.String(200))
+
+    def add(self):
+        """ Adds the current values of the Users fields to SQLAlchemy session and then writes the changes to the database """
+        
+        db.session.add(self)
+        db.session.commit()
+
+    @classmethod
+    def is_jti_blacklisted(cls, jti):
```
Comment (DannyLee12) [inline Analytics/models/revoked_tokens.py]: typehints?

File: Analytics/models/users.py [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+        self.timestamp = timestamp
+
+    def __repr__(self) -> str:
+        """
+        overrides the dunder repr method
+        :return: the users email address
+        """
+        return 'Users email is: %s' % self.email
+
+    def __str__(self) -> str:
```
Comment (DannyLee12) [inline Analytics/models/users.py]: This calls self.json(), which returns a dict. This should return a str.

File: Analytics/models/users.py [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+        """
+        overrides the dunder string method to cast user to a string
+        :return: a JSON string of the Users objects attributes
+        """
+        return self.json()
```
Comment (DannyLee12) [inline Analytics/models/users.py]:
```suggestion
        return json.dumps(self.json())
```

File: Analytics/models/users.py [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+    def save(self) -> NoReturn:
+        """
+        Add the current Users fields to the SQLAlchemy session
+        """
+        try:
+            db.session.add(self)
+            db.session.flush()
+        except IntegrityError as ie:
+            db.session.rollback()
+            print(self.fullname, 'User already exists')
```
Comment (DannyLee12) [inline Analytics/models/users.py]: logger, not print. please.

File: Analytics/models/users.py [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+        except IntegrityError as ie:
+            db.session.rollback()
+
+    @staticmethod
+    def commit() -> NoReturn:
+        """ Commit updated items to the database """
+        db.session.commit()
+
+    @classmethod
+    def find_by_email(cls, email) -> object:
```
Comment (DannyLee12) [inline Analytics/models/users.py]: typehint
```suggestion
    def find_by_email(cls, email: str) -> object:
```

File: Analytics/models/users.py [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+        Return the user corresponding to the email argument from the Users table
+        :param email: user's email
+        :type email: string
+        :return: Users credentials
+        :rtype: Users instance
+        """
+        return cls.query.filter_by(email=email).first()
+
+    @staticmethod
+    def generate_hash(password) -> bytes:
```
Comment (DannyLee12) [inline Analytics/models/users.py]:
```suggestion
    def generate_hash(password: str) -> bytes:
```

File: Analytics/models/users.py [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+        Generate a secure hash
+        :param password: user's to be password
+        :type password: string
+        :return: a hashed version of the password string
+        :rtype: unicode encoded string
+        """
+        return bcrypt.hashpw(password, bcrypt.gensalt())
+
+    @staticmethod
+    def verify_hash(password, hash) -> bool:
```
Comment (DannyLee12) [inline Analytics/models/users.py]:
```suggestion
    def verify_hash(password: str, hash: bytes) -> bool:
```

File: Analytics/models/widget.py [hunk 30]
```diff
@@ -0,0 +1,145 @@
...
+        """
+        Check if table exists
+        :return: True if the table exists in the database otherwise False
+        """
+        # Does the table exist?
+        has_table = db.engine.dialect.has_table(db.engine, self._WIDGET_DB_TABLE_NAME)
+        return has_table
+
+    @classmethod
+    def get_widget_by_id(cls, widgetID) -> object:
```
Comment (DannyLee12) [inline Analytics/models/widget.py]:
```suggestion
    def get_widget_by_id(cls, widgetID: int) -> object:
```

Comment (DannyLee12) [reply Analytics/models/widget.py]: Object?

File: Analytics/models/widget.py [hunk 31]
```diff
@@ -0,0 +1,145 @@
...
+        """
+        Fetches a widget by its id
```
Comment (DannyLee12) [inline Analytics/models/widget.py]: Fetches the first instance of a widget matching the widget id.

I hope that this is the desired behavior.

File: Analytics/models/widget_layout.py [hunk 32]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        Layouts initialise
+        :param widget_id: widgets identification number the layout is related to
+        :param x: x coordinate
+        :param y: y coordinate
+        :param height: height
+        :param width:   width
+        :param static:  true if the layout is static and cannot be moved
+
+        :type widget_id: int
```
Comment (DannyLee12) [inline Analytics/models/widget_layout.py]: These aren't python3 typehints.

File: Analytics/models/widget_layout.py [hunk 33]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        self.widget_id = widget_id
+        self.x_coord = x
+        self.y_coord = y
+        self.height = height
+        self.width = width
+        self.static = static
+
+        # Check if tables exsists
+        # self.create_table()
```
Comment (DannyLee12) [inline Analytics/models/widget_layout.py]: remove this

File: Analytics/models/widget_layout.py [hunk 34]
```diff
@@ -0,0 +1,161 @@
...
+        super().__init__()
+
+    def __str__(self) -> str:
+        """
+        Overides the dunder string method an returns the layout as a JSON
+        :return: a json as a string
+        """
+        return self.json()
```
Comment (DannyLee12) [inline Analytics/models/widget_layout.py]: This doesn't return a string?

File: Analytics/resources/Widgets/create_widget_layout.py [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+    :param static: layout static property
+
+    :type widgetID: String
+    :type x: Integer
+    :type y: Integer
+    :type h: Integer
+    :type w: Integer
+    :type static: String
+
+    :returns:  on success a HTTP status code 204, executed successful with no content. otherwise
```
Comment (DannyLee12) [inline Analytics/resources/Widgets/create_widget_layout.py]: 204 is a deleted status code.

File: Analytics/resources/Widgets/create_widget_layout.py [hunk 36]
```diff
@@ -0,0 +1,111 @@
...
+        self.post_reqparser.add_argument("x", help='Widget layout: x coordinate found', location=['form', 'json'])
+        self.post_reqparser.add_argument("y", help='Widget layout: y coordinate found', location=['form', 'json'])
+        self.post_reqparser.add_argument("h", help='Widget layout: height found', location=['form', 'json'])
+        self.post_reqparser.add_argument("w", help='Widget layout: width found', location=['form', 'json'])
+        self.post_reqparser.add_argument("static", default=False, help='Widget layout: error passing static variable',
+                                         location=['form', 'json'])
+        super().__init__()
+
+
+    def post(self) -> tuple:
```
Comment (DannyLee12) [inline Analytics/resources/Widgets/create_widget_layout.py]:
```suggestion
    def post(self) -> message: str, status_code: int
```

Comment (DannyLee12) [reply Analytics/resources/Widgets/create_widget_layout.py]: Perhaps the brackets are required.

File: Analytics/resources/Widgets/get_layouts.py [hunk 37]
```diff
@@ -0,0 +1,70 @@
...
+        """
+        self.reqparser = reqparse.RequestParser()
+        self.reqparser.add_argument('userID', required=True, help='A userID is required',
+                                    location=['form', 'json'])
+        self.reqparser.add_argument('limit', required=False, default=10, help='A userID is required',
+                                    location=['form', 'json'])
+        super().__init__()
+
+    @jwt_required
+    def post(self) -> tuple:
```
Comment (DannyLee12) [inline Analytics/resources/Widgets/get_layouts.py]: What is in the tuple

File: Analytics/resources/Widgets/get_widget_layout.py [hunk 38]
```diff
@@ -0,0 +1,59 @@
...
+        :type widgetID:   Integer
+        """
+        # Arguments required to fetch the layout the widget related to the userID
+        self.reqparser_get = reqparse.RequestParser()
+        self.reqparser_get.add_argument('widgetID', required=True, help='A widgetID is required',
+                                        location=['form', 'json'])
+        super().__init__()
+
+    @jwt_required
+    def post(self) -> tuple:
```
Comment (DannyLee12) [inline Analytics/resources/Widgets/get_widget_layout.py]: Please put types

File: Analytics/resources/Widgets/get_widgets.py [hunk 39]
```diff
@@ -0,0 +1,70 @@
...
+        # Arguments required to fetch the widgets related to the userID
+        self.reqparser_get = reqparse.RequestParser()
+        self.reqparser_get.add_argument('userID', required=True, help='A user_id is required',
+                                        location=['form', 'json'])
+        self.reqparser_get.add_argument('limit', default=1, required=False, type=int,
+                                        help='unable to parse limit', location=['form', 'json'])
+        super().__init__()
+
+    @jwt_required
+    def post(self) -> tuple:
```
Comment (DannyLee12) [inline Analytics/resources/Widgets/get_widgets.py]: Types

**Pull request merged**